feat/fix: Add support for wget2 and aria2c and fix spelling typos

lainedfles opened the pull request:
This update introduces the ability to use wget2 or aria2c for downloading ZIM files. Users can now choose between different downloaders by specifying `-2` for `wget2` or `-3` followed by max connections for `aria2c`. Additionally, the script now checks for the existence of these commands and includes improved progress tracking and parallel download capabilities.

Changes include:
- Added new command-line options for selecting downloaders
- Introduced support for parallel downloads using aria2c
- Enhanced error handling for missing commands
- Correct spelling typo errors

Commit:
feat/fix: Add support for wget2 and aria2c and fix spelling typos

This update introduces the ability to use wget2 or aria2c for downloading ZIM files. Users can now choose between different downloaders by specifying `-2` for `wget2` or `-3` followed by max
connections for `aria2c`. Additionally, the script now checks for the existence of these commands and includes improved progress tracking and parallel download capabilities.

Changes include:
- Added new command-line options for selecting downloaders
- Introduced support for parallel downloads using aria2c
- Enhanced error handling for missing commands
- Correct spelling typo errors

lainedfles commented:
Resolves #38

jojo2357 commented on kiwix-zim-updater.sh:
```diff
@@ -526,7 +586,7 @@ for ((i = 0; i < ${#LocalZIMNameArray[@]}; i++)); do
       else
         echo "    Checking for online checksum..."
-        if wget -S --spider -q -O - "$BaseURL$MatchingCategory/$FileName.sha256" >/dev/null 2>&1; then
+        if $WGET_CMD -S --spider -q -O - "$BaseURL$MatchingCategory/$FileName.sha256" >/dev/null 2>&1; then
```
Should this check `if [[ -n "$ARIA2C_CMD" ]]`?

lainedfles replied:
I don't think so, aria2c isn't used here, only wget. Using a variable here is necessary for those of us on distributions (like Arch Linux) with wget2 packaged as `wget2`.

jojo2357 replied:
Oh i see,

> Enable use of `aria2c`, aria2c is only used for parallel downloads, wget is still required for everything else like metadata, etc

jojo2357 commented on kiwix-zim-updater.sh:
```diff
@@ -710,12 +770,20 @@ if [ $AnyDownloads -eq 1 ]; then
+            [[ $DEBUG -eq 0 ]] && $ARIA2C_CMD --summary-interval=$ARIA2C_SUMMARY_INTERVAL -x $ARIA2C_MAX_CONN -c -d "${FilePath%/*}" -o "${FilePath##*/}" "$DownloadURL" 2>&1 |& tee -a download.log # Download new ZIM
+	  else
+            [[ $DEBUG -eq 0 ]] && $WGET_CMD -q $WGET_SHOW_PROGRESS --progress=bar:force -c -O "$FilePath" "$DownloadURL" 2>&1 |& tee -a download.log # Download new ZIM
+	  fi
```
Could we just abstract this big thing to a function somehow?

lainedfles replied:
Sure, take a look at 878d8fd. Seems to work as expected.

Commit:
refactor: Add run_downloader function and remove duplicate download logic

- Added a new `run_ downloader` function to centralize and streamline the download process, supporting both `aria2c` and `wget`.
- Removed redundant code blocks that previously handled downloading in multiple places, reducing duplication.
- Improved maintainability by consolidating download logic into a single function.

jojo2357 commented:
Can you update the readme to reflect the changes to usage?

jojo2357 commented:
I get the following error when i try to run with `wget2` installed (alongside `wget`)

```
./kiwix-zim.sh ../kiwix-zim/fakezims/ -2
2. Preprocessing...

  -Validating ZIM directory...
  ✓ Valid ZIM Directory 

  -Building online ZIM list...
Unknown option 'show-progress'
    ✗  Could not find any remote files, exiting
```

lainedfles commented:
> I get the following error when i try to run with `wget2` installed (alongside `wget`)

Looks like a failure of this [condition](https://github.com/jojo2357/kiwix-zim-updater/pull/44/files#diff-33ed2d77f893ef40410e97365f5ca89a0449afd1c22928df7686e0c7ffc3172fR466-R470): `if $WGET_CMD -V | head -1 | grep -i "${WGET_CMD}\s\+2" &>/dev/null;`

What does your `wget2 -V` version string look like?

```
[user@host ~]$ wget2 -V
GNU Wget2 2.2.0 - multithreaded metalink/file/website downloader

+digest +https +ssl/gnutls +ipv6 +iri +large-file +nls -ntlm -opie +psl -hsts
+iconv +idn2 +zlib -lzma +brotlidec +zstd -bzip2 -lzip +http2 +gpgme

Copyright (C) 2012-2015 Tim Ruehsen
Copyright (C) 2015-2024 Free Software Foundation, Inc.

License GPLv3+: GNU GPL version 3 or later
<http://www.gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Please send bug reports and questions to <bug-wget@gnu.org>.

[user@host ~]$ command -v wget2
/usr/bin/wget2
```

Example:
```
[user@host ~]$ wget2 -V | head -1 | grep -i "${WGET_CMD}\s\+2"
GNU Wget2 2.2.0 - multithreaded metalink/file/website downloader

[user@host ~]$ wget2 -V | head -1 | grep -i "${WGET_CMD}\s\+2" &>/dev/null; echo $?
0
```

jojo2357 commented:
```
$ wget2 -V
GNU Wget2 1.99.1 - multithreaded metalink/file/website downloader

+digest +https +ssl/gnutls +ipv6 +iri +large-file +nls -ntlm -opie +psl +iconv +idn2 +zlib +lzma +brotlidec +bzip2 +http2 +gpgme
```

```
$ wget2 -V | head -1 | grep -i "${WGET_CMD}\s\+2" &>/dev/null; echo $?
1
```

lainedfles commented:
`1.99.1` is marked as a [6 year old alpha release](https://gitlab.com/gnuwget/wget2/-/releases/wget2-1.99.1). I'm guessing you're running some LTS distribution? Can you try `wget2 -V | head -1 | grep -i "${WGET_CMD}\s\+\(2\|1\.99\)"`?

jojo2357 commented:
That works as intended

Commit:
fix: Support wget2 1.99 alpha version string

- Recognize 1.99 wget2 version strings
- Update README to include usage changes

jojo2357 commented:
I tried running same as above, and got this error:

```
  Processing wikipedia_en_100_maxi_2023-05.zim
    ✓ Status : ZIM doesn't exist on disk. Downloading...

    Download (mirror) : https://dumps.wikimedia.org/kiwix/zim/wikipedia/wikipedia_en_100_maxi_2025-03.zim
Unknown progress type 'bar:force'
    Old : wikipedia_en_100_maxi_2023-05.zim
    New : wikipedia_en_100_maxi_2025-03.zim
    ✗ Status : New ZIM failed verification. Old ZIM not purged.
```

probably same cause as above if i had to guess

Commit:
fix: improve wget1/wget2 compatibility for progress display

Summary:
- Updated WGET_SHOW_PROGRESS for wget2 compatibility
- Removed conflicting --progress=bar:force option in specific commands
- Added conditional handling to ensure proper progress bar display across both wget versions

lainedfles commented:
I've made an attempt to extend compatibility for wget2 alpha and beta versions by omitting the `--progress=bar:force` option. Wouldn't it be better to upgrade wget2 to a modern version? It looks like wget2 version 2+ all recognize the `--progress=bar:force` and variant options.

jojo2357 commented:
my idea was to just check for ancient versions and disable progress bar iff `wget2` is the weird alpha.

apt's [official packages](https://launchpad.net/ubuntu/+source/wget2) ship `wget2 v1.99` as recently as 22.04 LTS. Seeing how that is still supported, I think it would be valid to support it.

lainedfles commented:
Great, hopefully the latest commit has achieved that goal. I'll try to set up a 1.99 testing environment.

Commit:
fix: improve wget2 compatibility by adding progress bar option

Add the --progress=bar option to maintain consistent progress bar behavior when using wget2, especially in alpha and beta versions.

lainedfles commented:
One more fix to restore the v1.99 progress bar. It seems that I already had a working 22.04 LTS VM where it seems to work as expected.

Commit:
fix: adjust pattern to match any form of progress:bar

jojo2357 commented:
I don't get a progress bar when downloading the index with `wget2` but I do get it during the download phase

Commit:
fix: progress arguments while downloading index

lainedfles commented:
Does the latest commit improve the behavior you see?

jojo2357 commented:
Thanks for getting back to me, I do not get a progress bar during the fetching of the index. I do still get the progress bar for the regular download.

If I had to guess I think it might have to do with how the command is being piped?

Commit:
fix: Work around wget2 progress bar bug with temporary file in master_scrape

This commit addresses a wget2 progress bar bug that prevents display during index fetching. A temporary file is introduced to stage the downloaded data, allowing reliable parsing
and avoiding stream corruption from progress output. Ensure logging to download.log for debugging and activate proper cleanup of the temp file after processing.

lainedfles commented:
Forgive me, I'd overlooked the output of this command entirely. Unlike `wget`, `wget2` (any version I'd tested including the 1.99 alpha) outputs the progress bar to standard output instead of standard error (**bug?**). This prevents the ability to perform any redirection and ends up writing progress output erroneously into `kiwix-index`. The latest commit re-structures the index handling to use a temporary file as a workaround.

Commit:
fix: Improve temporary file handling with mktemp check and add local ZIM size info

This commit addresses two main changes in the `kiwix-zim-updater.sh` script. First, it ensures proper temporary file creation by checking for the availability of `mktemp` before using it, falling
back to a default path if unavailable. This prevents errors on systems without `mktemp`, improving reliability. Second, it adds human-readable file size display (e.g., "100MiB") next to local ZIM
filenames during checks, enhancing user feedback and clarity.

jojo2357 commented:
I know this is really minor, and we are basically there (tysm btw), but would it be possible to change the progress display for regular `wget` to not show the temp file? 

current:
```
  -Building online ZIM list...
/tmp/tmp.Rj2uOWV948 100%[===================>]   3.80M  3.39MB/s    in 1.1s 
```

proposed:
```
  -Building online ZIM list...
                    100%[===================>]   3.80M  3.39MB/s    in 1.1s 
```

Perhaps we just save to `kiwix-index` with `-O` and then postprocess in the same way you've already done. This would show `kiwix-index` as the file being saved and would remove the weak dep of `mktemp`

Commit:
fix: Address wget2 progress bar issue with version-specific index fetching

This commit resolves a wget2 progress bar bug by introducing version-dependent logic for downloading the Kiwix catalog. When using wget2, a temporary file is now used to work around display issues,
while older wget versions use process substitution for direct output handling. The RawLibrary data processing and timestamp extraction are maintained across both workflows to ensure consistent index
updates.

lainedfles commented:
The index download for wget 1 now uses the previous method. I've removed the mktemp dependency and instead use `kiwix-index-dl` for a friendly name; this avoids potentially corrupting the index. I've granted maintainer access to the PR so feel free make changes.

lainedfles commented:
@simon-bueler Can you share more about your setup?

- Which distribution are you using?
- Do `wget` and `wget2` coexist on your distribution? Or does one replace the other?
- How do you execute the script?
- Does the output or log provide any insight?
- Are you using the script from the latest version of this pull request?

I've tested from **Arch Linux** and **Ubuntu 24.04** and confirmed that it operates correctly:
```bash
time ./kiwix-zim-updater.sh -2 -u -d -w -c /storage/kiwix
```

The output looks like:
```
. Downloading New ZIM(s)...

  Processing based.cooking_en_all_2024-04.zim
    ✓ Status : ZIM doesn't exist on disk. Downloading...

    Download (mirror) : https://mirror.download.kiwix.org/zim/zimit/based.cooking_en_all_2025-04.zim
/storage/kiwix/based 100% [====================>]   15.04M    4.96MB/s
                          [Files: 1  Bytes: 15.0]
    Calculating checksum for : /storage/kiwix/based.cooking_en_all_2025-04.zim
    ✓ Checksum passed

    Old : based.cooking_en_all_2024-04.zim
    New : based.cooking_en_all_2025-04.zim
    ✓ Status : New ZIM downloaded successfully. Old ZIM purged.
```

The latest revision of the script has following checksum:
```bash
sha256sum kiwix-zim-updater.sh
b6aaf668b720faf2e0d51a9dca13324ec8a9f64164eadc1bcbbd7a259c21b3c6  kiwix-zim-updater.sh
```

Commit:
fix: Update ZIM file path handling to use full path

1. Adds `-P` option to wget command(s) in case the download somehow resolves to a list
2. Updates variable references from `$NewZIM` to `$NewZIMPath` when
   checking for existing files to ensure that the full path is used

lainedfles commented:
@simon-bueler I believe that I've identified the problem you've encountered. The existing file check was only looking for the `$NewZIM` file in the relative path and not the fully declared path within `$NewZIMPath`. If there exist any matching files within your current working directory, the download would be skipped and the file move would fail. 47576f6 aims to resolve this. Could you test again with the latest version?

@JacksonChen666, I've attempted to add some basic checks and documentation regarding the bash/grep requirements for MacOS. The script shebang now uses `/usr/bin/env` so that you can set your `PATH` as desired and still execute with `./`. See 1009978 and 217cf2e. **Note** that I don't use MacOS nor have access to testing resources so hopefully you can validate. Relates to #45.

@jojo2357 We can split this PR if you would like to separate the referenced MacOS changes.

jojo2357 commented:
Agh sorry for ghosting you. Life stuff.

> We can split this PR if you would like to separate the referenced MacOS changes

That would be ideal.

Let me know when this PR is ready to go and ill re-review so we can knock this out

lainedfles commented:
@jojo2357 No problem, hopefully all is well? I think it's ready to go.

jojo2357 commented:
Took a look, i like the addition of the file size. 

When the script is `ctrl` + `c`'d while downloading with `wget`, the whole script terminates. With `wget2`, only the wget dies and the script continues. Moreover, the script continues processing including removing the old zim...even if the new file is 0 bytes. 

Thats the last thing *i* can see.

lainedfles commented:
On second thought, I don't want to spend any more time on this. I don't mind maintaining any changes I need locally. Feel free to use any of the changes without attribution if desired (I don't care).

jojo2357 commented:
Im sorry :pensive:

lainedfles commented:
Good luck!
